services/supervisor/src/transactions.rs

```diff
@@ -280,8 +328,8 @@ impl SupervisorInterface for Supervisor {
             return Err(Error::AlreadyDeployed.into());
         }
 
-        let confirmations = deploy_requests.confirm(&deploy, author);
-        if confirmations == deploy_requests.validators_amount() {
+        deploy_requests.confirm(&deploy, author);
+        if Mode::deploy_approved(&deploy, &deploy_requests) {
```

Comment:
Well, the original idea here was to perform some basic check that every validator at least **knows what should be deployed** before starting the actual deploy process (so if, e.g, N - 1 validators have the artifact and 1 does not, the deploy won't start, since it's sentenced to fail anyway).

Regarding docs: yeah, I believe that it should be noted directly, but I guess it's a topic for the different PR (which will expand the `Supervisor` docs, since they aren't really rich now).

Reply:
> so if, e.g, N - 1 validators have the artifact and 1 does not, the deploy won't start

But this requires efforts from the *admins* to of the network to perform actions, rather than nodes, doesn't it? As far as I understand, there is no workflow to ensure that artifact is downloaded to perform `DeployRequest` (and I'm not sure we should require one); deployment is actually performed by the *nodes* in `after_commit` after the request is being registered as pending.

> since it's sentenced to fail anyway

Why though?

Reply:
> But this requires efforts from the admins to of the network to perform actions, rather than nodes, doesn't it?

Well, I don't want to create an impression that I'm standing for it, I've just explained the original idea :)
In my opinion, we can change this behaviour if it will be considered reasonable.

> Why though?

As we discussed some time ago, there is a plan to support a mode when node that wasn't able to deploy a service will be (at least temporarily) disconnected from the network. Currently there is no such a mechanism, and thus the relationship of the network and the fail-prone node are not clear: the node will still **attempt** to be a part of the network, but actually it won't be able to be a valid network part (especially when it'll come to the inter-service communication).

When we'll have a mechanism for nodes to disable themselves upon failure, this restriction can (and I believe will) be lowered.

But now I think it's reasonable to keep it just to avoid ambiguity.

If "why" is related to the concrete code than to the ideology, [here's the implementation of this check](https://github.com/exonum/exonum/blob/ef00a439d0f3cda1e344efd61c12d67369d91412/services/supervisor/src/transactions.rs#L374).

Reply:
The code you've highlighted corresponds to processing of `DeployConfirmation`s, isn't it? I'm not arguing to change anything in their processing; IMO, it is completely reasonable to require all validator *nodes* to confirm artifact deployment. I'm arguing that processing of `DeployRequest`s may be modified for the decentralized supervisor (although this is not a critical issue).

Re: "disabling on failure": there is *already* point in time (namely, [`Dispatcher::commit_block`](https://github.com/exonum/exonum/blob/ef00a439d0f3cda1e344efd61c12d67369d91412/exonum/src/runtime/dispatcher/mod.rs#L265) in a block with an artifact queued via `start_artifact_registration()`) when deployment becomes blocking; a node does not continue working until it deployed the artifact successfully (at least, that is the intention of the code, it should be tested), and the node panics if it cannot deploy the artifact (which should be a permanent condition, but again this hasn't been tested in depth).

Reply:
>The code you've highlighted corresponds to processing of DeployConfirmations, isn't it?

Yep. To deploy an artifact, we need
1. Send N deploy requests (1 for simple supervisor, `validators_count` for decentralized).
2. Send `validators_count` (no matter if it's a simple or decentralized supervisor) deploy confirmations.

Deploy confirmations are sent after deploy is completed successfully (we need to distinguish deploy process and artifact registration, which are not equivalent: a deployed artifact will not be registered until all the deploy confirmations are collected).

So, at the very beginning, we can check that every service is able to *attempt* to deploy an artifact. If some service can't do it, then there will not be `validators_count` confirmations later, and deploy will fail. Thus, with that check we exit from the deployment cycle early.

*Again, that was the plan. Not sure if it works as intended though, and not sure if this plan is good.*

> a node does not continue working until it deployed the artifact successfully (at least, that is the intention of the code, it should be tested), and the node panics if it cannot deploy the artifact (which should be a permanent condition, but again this hasn't been tested in depth).

To be honest, I'm a bit lost when panics are caught and suppressed, and when they are actually behave like a panic. *Off-topic mumbling: when did we come to the point when panics are not always what they suppose to mean?...*

Even if deploy failure already results in the node disabling, it's still a whatever of "implementation detail"/"undocumented feature" -- nobody can know it until he'll dig into the sources of dispatcher. What I meant is a documented invariant which explicitly stated in several places, desirably I N  B I G L E T T E R S, just so nobody will be able to miss it.

By the way: even though node panics upon a failure, its key still persists in the network config and connect lists. I don't know if it's okay, but maybe node should shutdown gracefully, with all the disconnects and associated work before the actual application termination.

Reply:
By the way, currently if the node will be disabled, it's key still will be in the list of validator keys and thus deploy will fail, since there will be no confirmation from this node.
I don't know how to fix that easily. Maybe some farewell message like "I'm dead, guys, good luck to all of ya" will work.